src/Build.UnitTests/ChangeWaves_Tests.cs

```diff
@@ -0,0 +1,236 @@
+        [InlineData("16=7")]
+        public void InvalidCallerForIsFeatureEnabledThrows(string waveToCheck)
+        {
+            using (TestEnvironment env = TestEnvironment.Create())
+            {
+                env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", "16.8");
+                Shouldly.Should.Throw<InternalErrorException>(() => ChangeWaves.IsChangeWaveEnabled(waveToCheck));
+            }
+        }
+
+        [Theory]
+        [InlineData("test", "16.8")]
+        [InlineData("16_8", "5.7")]
+        [InlineData("16x8", "20.4")]
+        [InlineData("garbage", "18.20")]
+        public void InvalidFormatThrowsWarningAndLeavesFeaturesEnabled(string disableFromWave, string featureWave)
+        {
+            using (TestEnvironment env = TestEnvironment.Create())
+            {
+                env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", disableFromWave);
+                BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly();
+                ChangeWaves.IsChangeWaveEnabled(featureWave).ShouldBe(true);
+
+                string projectFile = @"
+                    <Project>
+                        <Target Name='HelloWorld' Condition=""'$(MSBUILDDISABLEFEATURESFROMVERSION)' == '" + ChangeWaves.EnableAllFeatures + @"' and $([MSBuild]::IsChangeWaveEnabled('" + featureWave + @"'))"">
+                            <Message Text='Hello World!'/>
+                        </Target>
+                    </Project>";
+
+                TransientTestFile file = env.CreateFile("proj.csproj", projectFile);
+
+                ProjectCollection collection = new ProjectCollection();
+                MockLogger log = new MockLogger();
+                collection.RegisterLogger(log);
+
+                Project p = collection.LoadProject(file.Path);
+                p.Build().ShouldBeTrue();
+
+                log.AssertLogContains("invalid format");
+                log.AssertLogContains("Hello World!");
+                ChangeWaves.DisabledWave = null;
+                BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly();
+            }
+        }
+
+        [Theory]
+        [InlineData("0.8")]
+        [InlineData("203.45")]
+        public void OutOfRotationWavesThrowsWarningAndDisablesFeatures(string disableFromWave)
+        {
+            using (TestEnvironment env = TestEnvironment.Create())
+            {
+                env.SetEnvironmentVariable("MSBUILDDISABLEFEATURESFROMVERSION", disableFromWave);
+                BuildEnvironmentHelper.ResetInstance_ForUnitTestsOnly();
+
+                string projectFile = @"
+                    <Project>
+                        <Target Name='HelloWorld' Condition=""'$(MSBUILDDISABLEFEATURESFROMVERSION)' == '" + ChangeWaves.LowestWave + @"' and $([MSBuild]::IsChangeWaveEnabled('" + ChangeWaves.LowestWave + @"')) == false"">
+                            <Message Text='Hello World!'/>
+                        </Target>
+                    </Project>";
```

Comment:
nit:
```suggestion
                string projectFile = $@"
                    <Project>
                        <Target Name='HelloWorld' Condition=""'$(MSBUILDDISABLEFEATURESFROMVERSION)' == '{ChangeWaves.LowestWave}' and $([MSBuild]::IsChangeWaveEnabled('{ChangeWaves.LowestWave}')) == false"">
                            <Message Text='Hello World!'/>
                        </Target>
                    </Project>";
```
And similar for other cases.

Reply:
how did I not do this by default 🤦